fsw/cfe-core/src/inc/cfe_msg.h

```diff
@@ -0,0 +1,220 @@
+/*
+**  GSC-18128-1, "Core Flight Executive Version 6.7"
+**
+**  Copyright (c) 2006-2019 United States Government as represented by
+**  the Administrator of the National Aeronautics and Space Administration.
+**  All Rights Reserved.
+**
+**  Licensed under the Apache License, Version 2.0 (the "License");
+**  you may not use this file except in compliance with the License.
+**  You may obtain a copy of the License at
+**
+**    http://www.apache.org/licenses/LICENSE-2.0
+**
+**  Unless required by applicable law or agreed to in writing, software
+**  distributed under the License is distributed on an "AS IS" BASIS,
+**  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+**  See the License for the specific language governing permissions and
+**  limitations under the License.
+*/
+
+/******************************************************************************
+ * Message access APIs
+ */
+
+#ifndef _cfe_msg_
+#define _cfe_msg_
+
+/*
+ * Includes
+ */
+#include "common_types.h"
+#include "cfe_mission_cfg.h"
+#include "cfe_sb.h"
+
+/** @defgroup CFEAPIMSGChecksum cFE Message Checksum Control APIs
+ * @{
+ */
+
+/*****************************************************************************/
+/**
+** \brief Gets the checksum field from a software bus message.
+**
+** \par Description
+**          This routine gets the checksum (or other message integrity check
+**          value) from a software bus message.  The contents and location of
+**          this field will depend on the underlying implementation of software
+**          bus messages.  It may be a checksum, a CRC, or some other algorithm.
+**          Users should not call this function as part of a message integrity
+**          check (call #CFE_SB_ValidateChecksum instead).
+**
+** \par Assumptions, External Events, and Notes:
+**          - If the underlying implementation of software bus messages does not
+**            include a checksum field, then this routine will return a zero.
+**
+** \param[in]  MsgPtr      A pointer to the buffer that contains the software bus message.
+**                         This must point to the first byte of the message header.
+**
+** \return The checksum included in the software bus message header (if present), otherwise,
+**         returns a checksum value of zero.
+**/
+uint16 CFE_MSG_GetChecksum(CFE_SB_MsgPtr_t MsgPtr);
```

Comment:
> If we are creating a new/separate API layer this shouldn't assume that the checksum is a uint16. What if its bigger (i.e. CRC32/MD5/SHA1)? What if the message does not even have a checksum? There is no opportunity for a descriptive return code here.
> 
> While I understand this is only a first take, it would be OK if this was confined as an internal header file but it is being exposed in the public API so if users actually start calling this stuff, we'll be in the same boat where the API is limiting the implementation options.

Higher-level topic--what are checksums for, from the CFE application's point of view? I can see there being value in comparing two messages to see if they are identical, and possibly reporting a checksum in a log or event for debugging purposes, but I think just "returning a checksum" may be inappropriate. Another option would be caller passes in a (large enough) buffer and can use memcmp to see if two sums are the same.

Reply:
Good conversation for future development, for now moved internal to msg (still planning some minor cleanup but it won't solve everything).  b5c8be6